workflow/controller/controller.go

```diff
@@ -848,6 +866,52 @@ func getWfPriority(obj interface{}) (int32, time.Time) {
 	return int32(priority), un.GetCreationTimestamp().Time
 }
 
+// 10 minutes in the past
+const maxCompletedStoreTime = time.Second * -600
+
+func (wfc *WorkflowController) cleanCompletedWorkflowsRecord() {
+	cutoff := time.Now().Add(maxCompletedStoreTime)
+	removeIndex := -1
+	wfc.recentCompletions.mutex.Lock()
+	defer wfc.recentCompletions.mutex.Unlock()
+
+	for i, val := range wfc.recentCompletions.completions {
+		if val.when.After(cutoff) {
+			removeIndex = i - 1
+			break
+		}
+	}
+	if removeIndex >= 0 {
+		wfc.recentCompletions.completions = wfc.recentCompletions.completions[removeIndex+1:]
+	}
+}
+
+func (wfc *WorkflowController) recordCompletedWorkflow(key string) {
+	if !wfc.checkRecentlyCompleted(key) {
+		wfc.recentCompletions.mutex.Lock()
+		wfc.recentCompletions.completions = append(wfc.recentCompletions.completions,
+			recentlyCompletedWorkflow{
+				key:  key,
+				when: time.Now(),
+			})
+		wfc.recentCompletions.mutex.Unlock()
+	}
+}
+
+func (wfc *WorkflowController) checkRecentlyCompleted(key string) bool {
+	wfc.cleanCompletedWorkflowsRecord()
+	recent := false
+	wfc.recentCompletions.mutex.Lock()
+	for _, val := range wfc.recentCompletions.completions {
+		if val.key == key {
+			recent = true
+			break
+		}
+	}
```

Comment:
thinking about how we have an O(n) operation for each workflow reconciliation. Maybe it's not a big deal. I guess n could be however many workflows would have completed in the last 10 minutes, which may not be terrible. I know you probably made it a slice because you can quickly delete the old ones which is nice. Theoretically, you could actually store a `map[string]completedWorfkflow`  alongside your slice in `recentCompletions`. 
When you delete from the slice, you have the key in your `recentlyCompletedWorkflow` to delete from the map.

Reply:
The slice can be mass deleted, which was why I did it that way. A map would require a one by one deletion, and insertion is more expensive.
I don't think keeping both is a good idea as we have most of the downsides of both in terms of costs. If we want to go to a map, I'd turn GC into a separate go background task which runs around the map handling a few at a time then releasing the mutex before grabbing it again. This is more complex, but perhaps better. Happy to switch if you think it's important.

Reply:
okay, thanks for the analysis. I'm fine with it as is.

Reply:
For what it's worth, I think this O(n) still should be plenty fast relative to the rest of workflows even for largish slices (say 1M)